### PR TITLE
Rich Command line output colour customization

### PR DESCRIPTION
**Pull request by naveen521kk**
In this PR I have changed the logging and also added an option on customizing colored output using a `rich.cfg` file. SEE the screenshot below.
example-output
Below is how the configuration works out
cli-config-rich

### CODE REVIEW

File: manim/logger/console.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+from rich.theme import Theme
+
+def parseTheme(fp):
+    with open(fp,'r') as f:
+        content=f.read()
+        theme=json.loads(content)
+    customTheme=Theme(theme)
+    return customTheme
+try:
+    customTheme =  parseTheme("rich.cfg")
```
[the remark this answers is not in the document]
Comment (naveen521kk) [reply manim/logger/console.py:12]: Could you tell me where is the path to the working directory is stored?

[the remark this answers is not in the document]
Comment (eulertour) [reply manim/logger/console.py:12]: I'll leave it up to @naveen521kk's discretion, but doing this without the try/except is actually considered antipattern in python [[1]](https://docs.quantifiedcode.com/python-anti-patterns/readability/asking_for_permission_instead_of_forgiveness_when_working_with_files.html) [[2](ctrl+f EAFP)](https://docs.python.org/3.8/glossary.html).

**Comment by naveen521kk** [comment]
@eulertour Can you tell me what should I change so that it is compatible with it! Like should I create a place in `manim.cfg` or something like that as I don't know what you have planned in `#98`.

File: manim/logger/console.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (naveen521kk) [reply manim/logger/console.py:12]: I will leave it just like that as eulertour said.

**Comment by eulertour** [comment]
Essentially create a [color] section or something similar in the file, and store the defaults there instead of in `rich.cfg`. The code or the docs [here](https://docs.python.org/3/library/configparser.html) will be the best references.

**Comment by naveen521kk** [comment]
I will work on it and make it compatible!

**Comment by naveen521kk** [comment]
Also, @eulertour is `manim.cfg` to be placed in the current working by the user right?

**Comment by leotrs** [comment]
> Also, @eulertour is `manim.cfg` to be placed in the current working by the user right?

Yes.

**Comment by naveen521kk** [comment]
I think 🤔 I did it. Could you please whether this work.
Type in `python -m man in.logger` to open the colour configuration. It will create `manim.cfg`. Also, where can I write some documentation for it?

**Comment by PgBiel** [comment]
> I think 🤔 I did it. Could you please whether this work.
> Type in `python -m manim.logger` to open the colour configuration. It will create `manim.cfg`. Also, where can I write some documentation for it?

`manim.cfg` doesn't exist yet. It is being implemented in #98. Once that is merged, you will have to adapt your code to the new system

**Comment by naveen521kk** [comment]
> 
> 
> > I think thinking I did it. Could you please whether this work.
> > Type in `python -m manim.logger` to open the colour configuration. It will create `manim.cfg`. Also, where can I write some documentation for it?
> 
> `manim.cfg` doesn't exist yet. It is being implemented in #98. Once that is merged, you will have to adapt your code to the new system

I had an idea of how `manim.cfg` would and have just implemented what I thought it would be. I will edit it according once `#98` is merged.

**Comment by naveen521kk** [comment]
Now #98 is merged what should Indo for it to be merged?

**Comment by leotrs** [comment]
Hi @naveen521kk. Thanks for following up. I have not read the code for this PR, but the steps to follow now are essentially the following. Let' say you want to change the code of the output (I don't know how to do this, this is purely hypothetical, but I hope these steps clarify the problem.)

1. First, you need to add a new section to the `default.cfg` file inside `manim/`. You can call it whatever you want, for example `[rich]`.
2. Under that section, you can define whatever config variables you want, for example `rich_output_color`.
3. Inside `config.py`, you need to modify `_parse_config`. The value of `rich_output_color` that was set in `default.cfg` will be available to you within that function as `config_parser['rich']['rich_output_color']`. Please note that **all** values of `config_parser` are strings. If you need other types, you must convert them from string (see documentation [here](https://docs.python.org/3/library/configparser.html)). Once you have read (and/or converted, modified, or parsed) the value of `config_parser['rich']['rich_output_color']`, you need to store it in `config`. Please use the same key, but without the section name. For example `config['rich_output_color'] = do_something(config_parser['rich']['rich_output_color'])`.
4. You must modify the logger configuration using the new config variables. Let's assume this is done in `logger.py`. In that file, you must import the configuration `from .config import config` and then use the config dictionary to set the logger configuration. For example `logging.basicConfig(color=config['rich_output_color'], ...)`.

All of the above are of course examples and you will have to change the things that need changing. In summary, you need to make the following changes.

`default.cfg`
```
...
[rich]
rich_output_color = <some_color>
...
```

`config.py`
```
def _parse_config(config_parser, args):
    ...
    config['rich_output_color'] = do_something(config_parser['rich']['rich_output_color'])
    ...
```

`logger.py`
```
from .config import config
`logging.basicConfig(color=config['rich_output_color'], ...)`
```

Please do not modify other parts of `default.cfg` or `config.py` that have nothing to do with the config options you are adding. Thanks! Hope that helps!

**Comment by naveen521kk** [comment]
So @leotrs  I tried for the new config parser and #98 I found a problem. I am trying out that rich coloured output PR. I tried
```sh
python -m manim.logger
```
To execute the `__main__.py` file in logger directory which worked before this update. But now whatever I do, argparse runs up. Is this indented? If yes how can I really execute that file? Also, 
```sh
python -m manim.anything
```
Just execute the argparse ever if there is no submodule. Looks like this need to be checked.

**Comment by naveen521kk** [comment]
Also, I just noticed that `config.py` depends on `logger.py`. So, I have created a new config parser. If you have a better idea than that please tell. It is ready for review.

File: manim/logger.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import logging
+import configparser
+
+from rich.logging import RichHandler
+from rich.console import Console
+from rich.theme import Theme
+#from .config import config
```
Comment (Aathish04) [inline manim/logger.py:7]: Remove this line if it is unnecessary.
```suggestion
```

File: manim/logger.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+#print(config)
+def parseTheme(fp):
+    config = configparser.ConfigParser()
+    config.read(fp)
+    theme=dict(config['log.color'])
```
Comment (Aathish04) [inline manim/logger.py:13]: Use the config-parser in `config.py` to do any parsing. 

Add the dictionary corresponding to `log.color` to `configparser.config` and/or `configparser.file_writer_config`  with the key `"log.color"` or something similar.

The `configparser.config` or `configparser.file_writer_config` dictionaries can be used for any config data that needs to be accessible by `logger`. 

Mentioning @leotrs for his thoughts.

File: manim/logger.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    customTheme=Theme(theme)
+    return customTheme
+try:
+    customTheme =  parseTheme("manim.cfg")
```
Comment (Aathish04) [inline manim/logger.py:17]: This is related to the comment immediately above this one.

`manim.cfg` in the current working directory won't be the only configuration file, as far as I can tell.

Either you can also read the themes present in all the other cascading `manim.cfg` files that eulertour mentioned well be implementing [here](https://github.com/ManimCommunity/manim/pull/98#issuecomment-653929650), or you can the config-parser in config.py to do any parsing.

File: setup.py [hunk 5]
```diff
@@ -11,7 +11,7 @@
         "console_scripts": [
             "manim=manim.__main__:main",
             "manimcm=manim.__main__:main",
-            "manim-logsetup=manim.logger.__main__:main",
+            "cfg-manim=manim.cfgwriter.__main__:main",
```
Comment (Aathish04) [inline setup.py:14]: OPINION:
`manim-cfg-setup` would be a clearer name for this command.

File: manim/logger.py (hunk 3 above)
Comment (naveen521kk) [reply manim/logger.py:13]: There was a problem while I was doing so. `config.py` imports this file while running and trying to import using your method creates an import error as the root module `manim` would has not been initiated and there would be an circular import.

File: manim/logger.py (hunk 2 above)
Comment (naveen521kk) [reply manim/logger.py:7]: definitely

File: manim/logger/console.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Aathish04) [reply manim/logger/console.py]: This will be changed once the rich theme writer is turned into a full-fledged manim cfg writer.

File: manim/logger.py (hunk 3 above)
Comment (Aathish04) [reply manim/logger.py:13]: This will be changed once the rich theme writer is turned into a full-fledged manim cfg writer.

File: manim/cfgwriter/__main__.py [hunk 6]
```diff
@@ -15,45 +15,60 @@ def check_valid_colour(color):
...
-if __name__=="__main__":
-    main()
\ No newline at end of file
+    console.print(
+        "A file called [yellow]manim.cfg[/yellow] is created. To save your theme please save that file and each time \nplace it in you current working directory(The directory where you execute manim command) for the your theme to be displayed."
+    )
+
+
+if __name__ == "__main__":
+    main()
```
Comment (Aathish04) [inline manim/cfgwriter/__main__.py:74]: This would be better off inside `utils` as a single script instead of a module.

File: manim/config_utils.py [hunk 7]
```diff
@@ -333,15 +359,17 @@ def _from_command_line():
...
 
 else:
     # In this case, we still need an empty args object.
     args = _parse_cli([], input=False)
     # Need to populate the options left out
-    args.file, args.scene_names, args.output_file = '', '', ''
+    args.file, args.scene_names, args.output_file = "", "", ""
 
 config_parser = configparser.ConfigParser()
 successfully_read_files = config_parser.read(config_files)
```
Comment (Aathish04) [inline manim/config_utils.py:375]: Since this has been converted into a utility, it should live in `utils` and not at the top level.

File: manim/logger.py [hunk 8]
```diff
@@ -4,21 +4,24 @@
 from rich.console import Console
 from rich.logging import RichHandler
 from rich.theme import Theme
-from .config_utils import config_parser,successfully_read_files
+
+from .config_utils import config_parser, successfully_read_files
+
 
 def parseTheme(fp):
```
Comment (Aathish04) [inline manim/logger.py:11]: Use snake_case please.

**Comment by Aathish04** [comment]
Mentioning @leotrs for his review.

File: manim/config.py [hunk 9]
```diff
@@ -1,38 +1,38 @@
-"""
-config.py
----------
-
-Process the manim.cfg file and the command line arguments into a single
-config object.
-
-"""
```
Comment (leotrs) [inline manim/config.py:8]: Why was this deleted? [PEP257](https://www.python.org/dev/peps/pep-0257/) clearly says

> All modules should normally have docstrings, and all functions and classes exported by a module should also have docstrings.

Please add it back.

File: setup.py [hunk 10]
```diff
@@ -11,6 +11,7 @@
         "console_scripts": [
             "manim=manim.__main__:main",
             "manimcm=manim.__main__:main",
+            "cfg-manim=manim.utils.cfgwriter:main",
```
Comment (leotrs) [inline setup.py:14]: I'd rather `manim-cfg`. @eulertour ?

File: manim/utils/config_utils.py [hunk 11]
```diff
@@ -0,0 +1,379 @@
+"""
+config.py
```
Comment (leotrs) [inline manim/utils/config_utils.py:2]: Change to `config_utils.py`

File: manim/utils/config_utils.py [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+---------
+
+Process the manim.cfg file and the command line arguments into a single
+config object.
```
Comment (leotrs) [inline manim/utils/config_utils.py:6]: Change to "utility functions for parsing manim config files."

File: manim/utils/config_utils.py [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+    args = _parse_cli([], input=False)
+    # Need to populate the options left out
+    args.file, args.scene_names, args.output_file = "", "", ""
+
+config_parser = configparser.ConfigParser()
+successfully_read_files = config_parser.read(config_files)
+
+
+# this is for internal use when writing output files
+file_writer_config = _parse_file_writer_config(config_parser, args)
```
Comment (leotrs) [inline manim/utils/config_utils.py:379]: It's fairly unorthodox to have code that executes at import inside a `utils` file. Must this go here?

File: manim/utils/cfgwriter.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    ----------
+    color : :class:`str`
+        The color to check whether it is valid.
+    Returns
+    -------
+    Boolean
+        Returns whether it is valid color or not.
+    """
+    try:
+        Color.parse(color) is None
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:23]: I don't understand what the `is None` clause is doing.

File: manim/utils/cfgwriter.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    except:
+        return False
+
+
+def main():
+    console = Console()
+    config = configparser.ConfigParser()
+    config.read(
+        successfully_read_files
+    )
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:34]: Turn this into a single line.

File: example_scenes/manim.cfg [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+logging.level.notset = dim
+logging.level.debug = yellow
+logging.level.info = dim purple
+logging.level.warning = dim red
+logging.level.error = red
+logging.level.critical = red
+log.level = 
+log.time = dim yellow
+log.message = green
+log.path = dim blue
```
Comment (leotrs) [inline example_scenes/manim.cfg:12]: Please add a newline at the end of file, as per PEP8

File: example_scenes/manim.cfg [hunk 17]
```diff
@@ -0,0 +1,12 @@
+[log.color]
```
Comment (leotrs) [inline example_scenes/manim.cfg:1]: I don't know how I feel about having a dot inside the name of a section. This is purely a matter of opinion, but I think this section should just be called `logger` or `rich`.

File: example_scenes/manim.cfg (hunk 16 above)
Comment (leotrs) [inline example_scenes/manim.cfg:12]: Similarly, the convention in the rest of the file is to use underscores between words, and not dots `.`. I assume these dots are necessary for rich purposes? If so, can we please consider using e.g. `log_message` here instead of `log.message` and then replacing those underscores when/if needed? Just a suggestion!

File: example_scenes/manim.cfg [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+logging.keyword = magenta
```
Comment (leotrs) [inline example_scenes/manim.cfg:2]: Please add a comment on this file that says that these colors are to be parsed by `rich` (or by colour?), and NOT the same colors that manim defines in `constants.py`.

File: manim/logger.py [hunk 19]
```diff
@@ -1,12 +1,46 @@
...
+    theme = dict(config_parser["log.color"])
+    customTheme = Theme(theme)
+    return customTheme
+
 
+try:
+    customTheme = parse_theme(successfully_read_files)
+    console = Console(theme=customTheme)
+except KeyError:
+    console = Console()
```
Comment (leotrs) [inline manim/logger.py:22]: Please print/log the error message here as well

File: manim/logger.py [hunk 20]
```diff
@@ -1,12 +1,46 @@
...
 
+# These keywords Are Highlighted specially.
+RichHandler.KEYWORDS = [
```
Comment (leotrs) [inline manim/logger.py:25]: I like this!

File: manim/utils/cfgwriter.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+        "logging.level.debug": "green",
+        "logging.level.info": "blue",
+        "logging.level.warning": "red",
+        "logging.level.error": "red bold",
+        "logging.level.critical": "red bold reverse",
+        "log.level": "",
+        "log.time": "cyan dim",
+        "log.message": "",
+        "log.path": "dim",
+    }
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:47]: There is no reason to define this again since these all are (or should all be) in the `default.cfg` file, under the logger section.

File: manim/utils/cfgwriter.py [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+        "[yellow bold]Manim Logger Configuration Editor[/yellow bold]", justify="center"
+    )
+    console.print(
+        "[red]The default colour is shown as input Statement.\nIf left empty default value will be assigned.[/red]"
+    )
+    console.print(
+        "[magenta]Please follow the link for available styles.[/magenta][link=https://rich.readthedocs.io/en/latest/style.html]docs[/link]"
+    )
+    for key in default:
+        console.print("Enter the Style for %s" % key + ":", style=key, end="")
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:58]: Does `rich` suggest the use of the `%` operator? If I remember correctly, `logger` asks you to pass the string as well as the interpolators as arguments, and not use the `%` operator.

File: manim/utils/cfgwriter.py [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+        temp = input()
+        if temp:
+            while check_valid_colour(temp) == False:
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:61]: Change to `while not check_valid_colour(temp)`

File: manim/utils/cfgwriter.py [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+                    "[red bold]Your Style is not valid. Try again.[/red bold]"
+                )
+                console.print("Enter the Style for %s" % key + ":", style=key, end="")
+                temp = input()
+            else:
+                default[key] = temp
+    config["log.color"] = default
+    for n in track(range(100), description="Converting to Manim.cfg"):
+        with open("manim.cfg", "w") as fp:
+            config.write(fp)
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:72]: I'd strongly suggest abstracting these lines into a separate function. Also, are we just writing to the current directory? We should change that, pending #179

File: manim/config.py (hunk 9 above)
Comment (naveen521kk) [reply manim/config.py:8]: I missed it out while copying. Will add :)

File: manim/utils/cfgwriter.py (hunk 22 above)
Comment (naveen521kk) [reply manim/utils/cfgwriter.py:58]: Well this is a normal print statement. So `%` operator is allowed.

File: manim/utils/cfgwriter.py (hunk 24 above)
Comment (Aathish04) [reply manim/utils/cfgwriter.py:72]: @leotrs With reference to the cascading config file system that was discussed, aren't we going to have multiple config files that have the capability to override each other? Why should this write _only_ to the places mentioned it #179 ?

I suggest an option could be added where the user can choose the path to save the config to.

File: manim/logger.py (hunk 19 above)
Comment (naveen521kk) [reply manim/logger.py:22]: Do we need to expicitely say people that you don't have a config file create one?

File: manim/utils/cfgwriter.py [hunk 25]
```diff
@@ -58,7 +63,7 @@ def main():
         console.print("Enter the Style for %s" % key + ":", style=key, end="")
         temp = input()
         if temp:
-            while check_valid_colour(temp) == False:
+            while check_valid_colour(temp):
```
Comment (Aathish04) [inline manim/utils/cfgwriter.py:66]: This line should be:
```suggestion
            while not check_valid_colour(temp):
```
The current while loop will run as long as the colour is valid. The suggested loop should run if the colour is invalid, which is the intended behaviour.

File: example_scenes/manim.cfg [hunk 26]
```diff
@@ -1,4 +1,5 @@
 [log.color]
+#It is parsed by Rich Colour Parse `rich.color`
```
Comment (Aathish04) [inline example_scenes/manim.cfg:2]:
```suggestion
# The colors here are parsed by rich.color , and are not related
# to the colours defined in constants.py's COLOR_MAP
```
I think this would be a bit clearer. The user may think that they will be able to input colours defined in Manim's own colour palette.

File: manim/logger.py (hunk 19 above)
Comment (leotrs) [reply manim/logger.py:28]: "No cfg file found, creating one in <path_to_file>" would be really good

File: manim/utils/cfgwriter.py (hunk 24 above)
Comment (leotrs) [reply manim/utils/cfgwriter.py:77]: At the moment, there are three different levels of config files. Library-wide, user-wide, and folder-wide. cfg-writer (nor anything/anyone) should change the first one (library-wide). When ran without arguments, then I think cfg/writer should change the second one (user-wide). If ran with some kind of flag/argument, then cfg-writer could change the third one (folder-wide).

All that just to say: I agree with you @Aathish04 :)

**Comment by naveen521kk** [comment]
@leotrs what else should be done for this PR to be merged?
- Merge Latest commits from master which I will do by tomorrow

File: manim/logger.py [hunk 27]
```diff
@@ -4,17 +4,23 @@
...
 def parse_theme(fp):
     config_parser.read(fp)
     theme = dict(config_parser["log.color"])
-    customTheme = Theme(theme)
+    try:
+        customTheme = Theme(theme)
+    except (color.ColorParseError,errors.StyleSyntaxError):
+        customTheme = None
+        printf("[logging.level.error]Looks like your colour configuration has an error. So loading default color configuration[/logging.level.error]")
+        #console.log(e) #Not logging it as it looks clumsy
```
Comment (leotrs) [inline manim/logger.py:21]: Please delete this line. In general, do not leave commented out lines in the codebase :)

**Comment by leotrs** [comment]
There are a bunch of open discussions on your latest commits. In particular, see my comment on executing code at import in config_utils.py. Is there an alternative way?

File: manim/logger.py (hunk 19 above)
Comment (naveen521kk) [reply manim/logger.py:28]: Wait, its `logger.py` file and not `cfwriter.py` file. Always I will write, this line `No cfg file found in <path_to_file>`.

File: manim/utils/config_utils.py [hunk 28]
```diff
@@ -0,0 +1,382 @@
...
+    # The third way involves using `python -m manim ...`.  In this
+    # case, the CLI arguments passed to manim do not include 'manim',
+    # 'manimcm', or even 'python'.  However, the -m flag will always
+    # be the first argument.
+    from_python_m = sys.argv[0] == "-m"
+
+    return from_cli_command or from_python_m
+
+
+def _run_config():
```
Comment (naveen521kk) [inline manim/utils/config_utils.py:348]: This function has to be executed twice, once in `logger.py`  and again while importing `config.py` any better idea tha that?
@leotrs

File: manim/utils/cfgwriter.py (hunk 24 above)
Comment (naveen521kk) [reply manim/utils/cfgwriter.py]: Done.

File: manim/utils/config_utils.py [hunk 29]
```diff
@@ -0,0 +1,404 @@
...
+    from_python_m = sys.argv[0] == "-m"
+
+    return from_cli_command or from_python_m
+
+
+def _paths_config_file():
+    library_wide = os.path.abspath(
+        os.path.join(os.path.dirname(__file__), "..", "default.cfg")
+    )
+    if sys.platform.startswith("linux"):
```
Comment (naveen521kk) [inline manim/utils/config_utils.py:352]: I haven't tested for linux. Please do check this.

File: manim/utils/cfgwriter.py [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+                )
+                console.print("Enter the Style for %s" % key + ":", style=key, end="")
+                temp = input()
+            else:
+                default[key] = temp
+    for key in default:
+        temp = default[key]
+        del default[key]
+        key = key.replace(".", "_")
+        default[key] = temp
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:73]: Issue (DRY): same as lines 52-56. Please turn these into a function.

File: manim/config.py [hunk 31]
```diff
@@ -1,38 +1,37 @@
...
-from .logger import logger
+
 from . import constants
+from .utils.config_utils import _run_config, _init_dirs, _from_command_line
+
+from .logger import logger
+from .utils.tex import TexTemplate, TexTemplateFromFile
 
-__all__ = ["file_writer_config", "config", "camera_config"]
+__all__ = ["config", "camera_config"]
```
Comment (leotrs) [inline manim/config.py:18]: question: why was `file_writer_config` removed? It needs to be exported as well.

File: manim/config.py [hunk 32]
```diff
@@ -81,346 +79,19 @@ def _parse_config(config_parser, args):
...
-    args = _parse_cli([], input=False)
-    # Need to populate the options left out
-    args.file, args.scene_names, args.output_file = '', '', ''
-
-config_parser = configparser.ConfigParser()
-successfully_read_files = config_parser.read(config_files)
-logger.info(f'Read configuration files: {successfully_read_files}')
-
-# this is for internal use when writing output files
-file_writer_config = _parse_file_writer_config(config_parser, args)
```
Comment (leotrs) [inline manim/config.py:420]: question: why was this removed?

File: example_scenes/manim.cfg [hunk 33]
```diff
@@ -0,0 +1,82 @@
+[CLI]
```
[the remark this answers is not in the document]
Comment (naveen521kk) [reply example_scenes/manim.cfg:1]: I just generated this using `cfg-manim` command and didn't notice that. Will fix.

[the remark this answers is not in the document]
Comment (naveen521kk) [reply example_scenes/manim.cfg:1]: I just generated this using `cfg-manim` command and didn't notice that. Will fix.

File: manim/config.py (hunk 32 above)
Comment (naveen521kk) [reply manim/config.py:420]: Do see line [92](https://github.com/ManimCommunity/manim/blob/561f9bdb3ff095ec48f449b9c058f12640b9d7b5/manim/config.py#L92). I defined it there.

File: manim/utils/config_utils.py [hunk 34]
```diff
@@ -0,0 +1,404 @@
...
+            if os.path.exists(args.config_file):
+                config_files.append(args.config_file)
+            else:
+                raise FileNotFoundError(f"Config file {args.config_file} doesn't exist")
+        else:
+            script_directory_file_config = os.path.join(
+                os.path.dirname(args.file), "manim.cfg"
+            )
+            if os.path.exists(script_directory_file_config):
+                config_files.append(script_directory_file_config)
```
[the remark this answers is not in the document]
Comment (naveen521kk) [reply manim/utils/config_utils.py:391]: OK, here comes a new question. Which `manim.cfg` should be taken at first? The one in the current working directory or the one the script is located? See below for an example.
I could do if I am in CWD and there is a `scene.py` I can do
```sh
manim scene.py
```
Now there would no problem reading the `manim.cfg` file there. But if I do 
```sh
manim /path/to/scene.py
```
Now, there is a dilemma of which cfg file to be used. The one in CWD or the one in which lives in the folder where `scene.py` is there. Pleace do help me with fixing this the I would be resolve this comment. @leotrs

File: manim/config.py (hunk 32 above)
Comment (leotrs) [reply manim/config.py:420]: You're right!

File: manim/utils/config_utils.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (Aathish04) [reply manim/utils/config_utils.py:391]: The python interpreter considers the place from which you executed the `python` command to be it's current working directory, even if you are running a file which is somewhere else.

In keeping with this tradition, I would suggest using the CWD's `.cfg` file, and not the one next to the script holding your animation.

[the remark this answers is not in the document]
Comment (leotrs) [reply manim/utils/config_utils.py:391]: I strongly think that it should be the one that is next to the scene script. Otherwise, running the same command on the same script could generate different videos. I believe @eulertour has said something similar before.

File: manim/config.py [hunk 35]
```diff
@@ -91,7 +89,7 @@ def _parse_config(config_parser, args):
 
 args, config_parser, file_writer_config, successfully_read_files = _run_config()
 if _from_command_line():
-    logger.info(f"Read configuration files: {successfully_read_files}")
+    logger.info(f"Read configuration files: {os.path.abspath(successfully_read_files[-1])}")
```
Comment (leotrs) [inline manim/config.py:92]: Praise: good catch here.

File: manim/utils/cfgwriter.py [hunk 36]
```diff
@@ -15,8 +15,10 @@
 from rich.style import Style
 from rich.errors import StyleSyntaxError
 
+INVALID_STYLE_MSG = "[red bold]Your Style is not valid. Try again.[/red bold]"
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:19]: Issue: this file has a bunch of functions that are not meant to be used by outside users of the file. For example, `replace_keys` and `is_valid_style`. For this purpose, please add the `__all__` variable at the top of the file specifying which variables/functions this file exports.

(In the case when `is_valid_style` becomes useful for an outside user, then it should be moved to a utils file instead. But let's keep it here for now.)

File: manim/utils/cfgwriter.py [hunk 37]
```diff
@@ -34,10 +36,36 @@ def check_valid_style(style):
         return False
 
 
+def replace_keys(default):
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:39]: Praise: good job here!

File: manim/utils/cfgwriter.py [hunk 38]
```diff
@@ -34,10 +36,36 @@ def check_valid_style(style):
...
+    return default
+
+
 def main():
-    successfully_read_files = _run_config()[-1]
+    config_items = _run_config()
+    successfully_read_files = config_items[-1]
     console = Console()
-    config = configparser.ConfigParser()
+    config = config_items[1]
```
Comment (leotrs) [inline manim/utils/cfgwriter.py]: Nitpick: it's more pythonic to do `_, config_parser, _, successfully_read_files = _run_config(). This is preferable since it explicitly says which return values of `_run_config` are being used and which are being discarded.

File: manim/utils/cfgwriter.py [hunk 39]
```diff
@@ -34,10 +36,36 @@ def check_valid_style(style):
...
     config.read(successfully_read_files)
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:69]: Issue: This is unnecessary, as per the previous line. The `config` returned by `_run_config` has already read and processed all the files in `successfully_read_files`.

File: manim/utils/cfgwriter.py [hunk 40]
```diff
@@ -34,10 +36,36 @@ def check_valid_style(style):
...
     default = config["logger"]
     console.print(
```
Comment (leotrs) [inline manim/utils/cfgwriter.py:71]: Question: Is there a reason why there are three consecutive calls to `console.print` instead of just one that prints the whole message at once?

Comment (leotrs) [reply manim/utils/cfgwriter.py:71]: Also, these messages could also be defined as constants at the top of the file, just like `INVALID_STYLE_MSG`.

**Comment by leotrs** [comment]
@naveen521kk Good job on the latest commits! One standing question is that I'd love to merge a test along with this as well. @huguesdevimeux ideas?

**Comment by huguesdevimeux** [comment]
> @naveen521kk Good job on the latest commits! One standing question is that I'd love to merge a test along with this as well. @huguesdevimeux ideas?

Well, I don’t know at all. 
I heard that there is/will be a PR to save logs. Maybe we will be able to build a test when this is merged.
(But To be very frank, I don’t think we should wait to have a test before merging this, as this is not crucial).

**Comment by leotrs** [comment]
> (But To be very frank, I don’t think we should wait to have a test before merging this, as this is not crucial).

Hmm fair, let's talk about tests in #201 then
